### ISSUE TITLE
Nested configuration or beans only used in the host Test class

### ISSUE THREAD

**Issue by hantsy** [quarkusio/quarkus#16767]
## Description

I created a sample project to experience Oidc API protection with Auth0, but when writing tests, I encountered a few issues. 

One is from writing tests with the usage of `@TestSecurity` with properties file users/roles,

https://github.com/hantsy/quarkus-sandbox/blob/master/oidc-api-auth0/src/test/java/com/example/TestSecurityLazyAuthTest.java

I tried to disable oidc in the Profile. 

https://github.com/hantsy/quarkus-sandbox/blob/master/oidc-api-auth0/src/test/java/com/example/PropertiesFileEmbeddedUsersProfile.java#L28

Then running my tests, it will report *OidcConfigurationMetadata* can not be injected. There is a `AudienceValiator` class which use the `OidcConfiguratoinMetadata`. 

https://github.com/hantsy/quarkus-sandbox/blob/master/oidc-api-auth0/src/main/java/com/example/security/AudienceValidator.java

I added a producer to expose mocked beans for them to overcome this barrier.

https://github.com/hantsy/quarkus-sandbox/blob/master/oidc-api-auth0/src/test/java/com/example/PropertiesFileEmbeddedUsersProfile.java#L42

But the producer is global in tests, and it will affect other tests.

## Implementation ideas

Spring Boot provides a simple mechanism to create test class scoped beans and configurations to override the default configurations and beans.

See: https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-testing-spring-boot-applications-excluding-config

It allow you create a temporary Configuration or Component via inner static  `@TestConfiguraiton` or inner static `@TestCompoent`.

```java 

@SpringBootTest
public class SomeTest{
  ...
  @TestConfiguration
  static class SomeConfig{
    @Bean
    SomeBean someBean(){}
    }
}
```

Hope there are some mechanisms in Quarkus to improve the testability.

eg. Make the specific **SomeBean** here only works for `SomeTest`, but not visiable for other tests.

```java 

@QuarkusTest
public class SomeTest{
  ...
  @TestDependent
  static class SomeProducers{
    @Produces
    SomeBean someBean(){}
    }
}
```

**Comment by mkouba** [comment]
> Spring Boot provides a simple mechanism to create test class scoped beans and configurations to override the default configurations and beans.

In Quarkus, you can try to use [testing profiles](https://quarkus.io/guides/getting-started-testing#testing_different_profiles) and enable an `@Alternative` bean for the specific profile.

**Comment by michalvavrik** [comment]
> In Quarkus, you can try to use [testing profiles](https://quarkus.io/guides/getting-started-testing#testing_different_profiles) and enable an @Alternative bean for the specific profile.

Right, I think that would work. Or I can see https://quarkus.io/guides/getting-started-testing#quarkus_mock that has some local test-class beans, which is what I think was the main point, to keep the bean visible within the test class. @holly-cummins @edeandrea is current support sufficient or do you plan some improvements? If so, maybe you want to add this to the https://github.com/orgs/quarkusio/projects/60 roadmap? Otherwise, let's close it.

As for described security issue, I can't understand where is the problem as links are invalid and there isn't a way to reproduce it. But it doesn't sound like security-per say, so I'll remove the label. If you disagree, please provide more context and I'll have a look. Thank you.

**Comment by edeandrea** [comment]
Mocking works very well. We do it all over the place in the Superheroes sample

see https://github.com/quarkusio/quarkus-super-heroes/blob/main/rest-narration/src/test/java/io/quarkus/sample/superheroes/narration/service/ImageGenerationServiceTests.java for one example using `@InjectMock`

There's also the [Alternative mechanism](https://quarkus.io/guides/getting-started-testing#mock-support) that @mkouba mentions.

You can also use [test profiles](https://quarkus.io/guides/getting-started-testing#testing_different_profiles) (see https://github.com/quarkusio/quarkus-super-heroes/blob/main/rest-fights/src/test/java/io/quarkus/sample/superheroes/fight/service/FightServiceConsumerContractTests.java for an example).
